### ISSUE TITLE
[Enhacement] Limit to only check 1 torrent at a time

### ISSUE THREAD

**Issue by latot** [arvidn/libtorrent#5897]
Hi hi :3

I found a little curious this, we can send to libtorrent check several torrent files, but, they are checked all in the same time...., we can change part of this with a limit of how many torrents can be active, but that will limit the seed and download too.

When we send several files to be checked at least I have 2 things that happens:

- With more torrents checking more slow it checks (we lose efficiency)
- The pc will be pretty locked until it finish

I think would be better, add an option to choose how much torrents can be checked at a time, or just limit it to 1.

Thx :D

**Comment by arvidn** [comment]
there's a separate limit for the number of checking torrents too (`active_checking`, [here](http://libtorrent.org/reference-Settings.html#active_checking)). But the limits only apply for `auto_managed` torrents, so you also have to make sure they have that flag set.

**Comment by arvidn** [comment]
The limit defaults to 1, so if you see multiple torrents check in parallel, chances are that you add torrents in "force-started" mode, i.e. without the auto_mangaged flag. Torrents are auto-managed by default as well though.

**Comment by latot** [comment]
Yes, this happens mainly when at least I force recheck the files..., can be this implemented to non-automangaged torrents?

**Comment by arvidn** [comment]
what's the behavior you're looking for?
If you want to queue torrents for checking, and only check *n* torrents at a time, that's what the auto-managed flag is for. Why do you want to be able to clear that flag while preserving this behavior?

**Comment by latot** [comment]
Maybe I'm confused with the auto-managed torrent.

A lot of times I download files in DD, and then I can share, or check the files if something is missed, but usually download in DD the files not always have the same names, or even have other content, so I need manually start checking the files with torrent, see the differences, fix names for example, run again to check.

If I have 2 torrents (usually more) to be checked, to avoid the torrent to start downloading all I need to pause them, so auto-managed need to be off, when I put to force recheck the torrent will start both torrents at the same time, and the computer is locked D:

I only can use auto-managed when the torrent is complete (all downloaded via DD) or need to fix/download missed things from it.

Obvs, I'll avoid do a lot of this manually, so I run a script that run force recheck in all incomplete torrents, there I notice all the files was checking at the same time.

There is the option to implement check one by one from the script, but I think this would be useful here.

**Comment by arvidn** [comment]
You can set the auto managed flag when you force-check the torrents, and then clear it again when they're done checking. To avoid the brief window between completing checking and you pausing the torrent, you can set the `upload_only` flag.

**Comment by latot** [comment]
Mmm, Is a very nice way to do it, just for an easy handling things, wouldn't be better set limits or params in manual torrents?

**Comment by arvidn** [comment]
> wouldn't be better set limits or params in manual torrents?

I don't understand what you propose

**Comment by latot** [comment]
To avoid handling the torrents in a script, when can be done here?

In a conceptual thing, maybe have not too much sense apply this considering that is "manual"...

**Comment by arvidn** [comment]
but how do you propose the behavior should be. Should the `active_checking` limit be special, and apply to both auto-managed and not auto-managed ones? There would be no way to exempt certain torrents from this limit in that case.

**Comment by latot** [comment]
In the particular case of checking files, I think have sense that the limit to be shared with manual, and auto-managed, just because we have a limit reading.

Now, just to have a "complete" answer in the sense to keep things "manual", I think this could be the behavior:
```active_checking_manual```:limit manual torrents to check 
```active_checking_auto```:limit torrents auto-managed 
```active_checking```: limit the plus of both

To simplify this, still is possible limit both with ```active_checking```

**Comment by arvidn** [comment]
I'm not convinced that *adding* two more limits makes this *simpler*. You can already do the same things with the knobs that exist. (at least that's my understanding).

The necessary documentation to describe the system would certainly be longer, as one metric of complexity.

**Comment by latot** [comment]
Mmm, yes, add two limits makes this "simpler", and "complex" because we need check relation betwen them...

Is true that we already have things to do this, is just a little complicated in some ways to handle over a script, maybe not that complicated, but the functions would be the same, for manual torrents check if something is checking, if not check the next, keep the process async ideally...

What about split both of them? ```active_checking_auto`` `and ```active_checking_manual```?

**Comment by arvidn** [comment]
> What about split both of them? active_checking_auto and active_checking_manual?

If going down that path, wouldn't it also then make sense to split all other limits into `*_auto` and `*_manual`? Or is `checking` special?

At that point, the `auto_managed` flag no longer means "apply limits". It becomes a selector of *which* limits to use. Why are there only two sets of limits to chose from? Perhaps there's a used case for 3 or more different sets of limits, and one should be able set up arbitrary limits.

In short, I think you'd need a strong argument why "2" is the correct magic number for sets of limits, or why "checking" is special from all other limits.

**Comment by latot** [comment]
That is right, At least I consider checking files a very special case.

When we set a limit to share or download is pretty simply how it works, we can download at 1MB/s, if we enable 2 torrents and the seeds have capacitiy, we will be able to download in total very close to 1MB/s, the same when we share.

When we check files the behavior es very different, if we check 1 torrent we will check at, whenever, 1MB/s (think is the limit to read), when we have 2 torrents checking we will check 0.8MB/s, (lets imagine 0.4 MB/s for each torrent), with more torrents more slow we will check (obvs, there is exceptions ,like, lets have 100 torrents of 100kb of every one), in the meanwhile we will lock the computer.

Global idea? just check what resource we use in every case, here a list with the status, is a very obvs list, is just to clarify:

On Hold:
-queued_for_checking

Download, low charge:
-downloading_metadata

Upload low charge:
-seeding

HD low charge:
-checking_resume_data

Nothing:
-finished

HD:
-checking_files
-allocating

Download High Charge:
-downloading

Obvs, the idea of the limits is can do a good use of the resources, if we will not use the pc we can use more resources, so, would be ideal in this case allocating and checking_files have the same limit for the use on the HD.

There is some balance about this, for example, doesn't affect too much Download on the HD just because the download speed actually is pretty low compared with the read/write speed of the HD, if the internet connection in the future increase to can affect I think would be nice check this again.

Even if we have ```auto``` and ```manual```, I think is pretty useful can have a limit on the resources to not overload it.

Maybe I write this.., too large, hopes i can wrote it well xD
